&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&ensp;
[Home Page](../../START_HERE.ipynb)



# Challenge - Gene Expression Classification


### Introduction

This notebook walks through an end-to-end GPU machine learning workflow where cuDF is used for processing the data and cuML is used to train machine learning models on it. 

After completing this excercise, you will be able to use cuDF to load data from disk, combine tables, scale features, use one-hote encoding and even write your own GPU kernels to efficiently transform feature columns. Additionaly you will learn how to pass this data to cuML, and how to train ML models on it. The trained model is saved and it will be used for prediction.

It is not required that the user is familiar with cuDF or cuML. Since our aim is to go from ETL to ML training, a detailed introduction is out of scope for this notebook. We recommend [Introduction to cuDF](../../CuDF/01-Intro_to_cuDF.ipynb) for additional information.

### Problem Statement:
We are trying to classify patients with acute myeloid leukemia (AML) and acute lymphoblastic leukemia (ALL) using machine learning (classification) algorithms. This dataset comes from a proof-of-concept study published in 1999 by Golub et al. It showed how new cases of cancer could be classified by gene expression monitoring (via DNA microarray) and thereby provided a general approach for identifying new cancer classes and assigning tumors to known classes. 

Here is the dataset link: https://www.kaggle.com/crawford/gene-expression.

## Here is the list of exercises and modules to work on in the lab:

- Convert the serial Pandas computations to CuDF operations.
- Utilize CuML to accelerate the machine learning models.
- Experiment with Dask to create a cluster and distribute the data and scale the operations.

You will start writing code from <a href='#dask1'>here</a>, but make sure you execute the data processing blocks to understand the dataset.



### 1. Data Processing

The first step is downloading the dataset and putting it in the data directory, for using in this tutorial. Download the dataset here, and place it in (host/data) folder. Now we will import the necessary libraries.

In [3]:
import numpy as np; print('NumPy Version:', np.__version__)
import pandas as pd
import sys
import sklearn; print('Scikit-Learn Version:', sklearn.__version__)
from sklearn import preprocessing 
from sklearn.utils import resample
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import SelectFromModel
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, roc_curve, auc
from sklearn.preprocessing import OrdinalEncoder, StandardScaler
import cudf
import cupy
# import for model building
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import mean_squared_error
from cuml.metrics.regression import r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn import linear_model
from sklearn.metrics import accuracy_score
from sklearn import model_selection, datasets
from cuml.dask.common import utils as dask_utils
from dask.distributed import Client, wait
from dask_cuda import LocalCUDACluster
import dask_cudf
from cuml.dask.ensemble import RandomForestClassifier as cumlDaskRF
from sklearn.ensemble import RandomForestClassifier as sklRF

NumPy Version: 1.20.3
Scikit-Learn Version: 0.24.2


We'll read the dataframe into y from the csv file, view its dimensions and observe the first 5 rows of the dataframe.

In [4]:
%%time
y = pd.read_csv('../../../data/actual.csv')
print(y.shape)
y.head()

(72, 2)
CPU times: user 2.61 ms, sys: 305 µs, total: 2.92 ms
Wall time: 2.49 ms


patient cancer
0        1    ALL
1        2    ALL
2        3    ALL
3        4    ALL
4        5    ALL

Let's convert our target variable categories to numbers.

In [5]:
y['cancer'].value_counts()
# Recode label to numeric
y = y.replace({'ALL':0,'AML':1})
labels = ['ALL', 'AML'] # for plotting convenience later on

Read the training and test data provided in the challenge from the data folder. View their dimensions.

In [6]:
# Import training data
df_train = pd.read_csv('../../../data/data_set_ALL_AML_train.csv')
print(df_train.shape)

# Import testing data
df_test = pd.read_csv('../../../data/data_set_ALL_AML_independent.csv')
print(df_test.shape)

(7129, 78)
(7129, 70)


Observe the first few rows of the train dataframe and the data format.

In [7]:
df_train.head()

Gene Description Gene Accession Number    1 call    2  \
0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -214    A -139   
1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -153    A  -73   
2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at  -58    A   -1   
3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at   88    A  283   
4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -295    A -264   

  call.1    3 call.2    4 call.3  ...   29 call.33   30 call.34   31 call.35  \
0      A  -76      A -135      A  ...   15       A -318       A  -32       A   
1      A  -49      A -114      A  ... -114       A -192       A  -49       A   
2      A -307      A  265      A  ...    2       A  -95       A   49       A   
3      A  309      A   12      A  ...  193       A  312       A  230       P   
4      A -376      A -419      A  ...  -51       A -139       A -367       A   

    32 call.36   33 call.37  
0 -124       A -135       A  
1  -79       A -186       A  
2  -37       A  -70       A  
3  330       A  337       A  
4 -188       A -407       A  

[5 rows x 78 columns]

Observe the first few rows of the test dataframe and the data format.

In [8]:
df_test.head()

Gene Description Gene Accession Number   39 call   40  \
0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -342    A  -87   
1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -200    A -248   
2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at   41    A  262   
3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at  328    A  295   
4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -224    A -226   

  call.1   42 call.2   47 call.3  ...   65 call.29   66 call.30   63 call.31  \
0      A   22      A -243      A  ...  -62       A  -58       A -161       A   
1      A -153      A -218      A  ... -198       A -217       A -215       A   
2      A   17      A -163      A  ...   -5       A   63       A  -46       A   
3      A  276      A  182      A  ...  141       A   95       A  146       A   
4      A -211      A -289      A  ... -256       A -191       A -172       A   

    64 call.32   62 call.33  
0  -48       A -176       A  
1 -531       A -284       A  
2 -124       A  -81       A  
3  431       A    9       A  
4 -496       A -294       A  

[5 rows x 70 columns]

As we can see, the data set has categorical values but only for the columns starting with "call". We won't use the columns having categorical values, but remove them.

In [9]:
# Remove "call" columns from training and testing data
train_to_keep = [col for col in df_train.columns if "call" not in col]
test_to_keep = [col for col in df_test.columns if "call" not in col]

X_train_tr = df_train[train_to_keep]
X_test_tr = df_test[test_to_keep]

Rename the columns and reindex for formatting purposes and ease in reading the data.

In [10]:
train_columns_titles = ['Gene Description', 'Gene Accession Number', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
       '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', 
       '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38']

X_train_tr = X_train_tr.reindex(columns=train_columns_titles)

In [11]:
test_columns_titles = ['Gene Description', 'Gene Accession Number','39', '40', '41', '42', '43', '44', '45', '46',
       '47', '48', '49', '50', '51', '52', '53',  '54', '55', '56', '57', '58', '59',
       '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71', '72']

X_test_tr = X_test_tr.reindex(columns=test_columns_titles)

We will take the transpose of the dataframe so that each row is a patient and each column is a gene.

In [12]:
X_train = X_train_tr.T
X_test = X_test_tr.T

print(X_train.shape) 
X_train.head()

(40, 7129)


0     \
Gene Description       AFFX-BioB-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-5_at   
1                                                     -214   
2                                                     -139   
3                                                      -76   

                                                      1     \
Gene Description       AFFX-BioB-M_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-M_at   
1                                                     -153   
2                                                      -73   
3                                                      -49   

                                                      2     \
Gene Description       AFFX-BioB-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioB-3_at   
1                                                      -58   
2                                                       -1   
3                                                     -307   

                                                      3     \
Gene Description       AFFX-BioC-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioC-5_at   
1                                                       88   
2                                                      283   
3                                                      309   

                                                      4     \
Gene Description       AFFX-BioC-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioC-3_at   
1                                                     -295   
2                                                     -264   
3                                                     -376   

                                                       5     \
Gene Description       AFFX-BioDn-5_at (endogenous control)   
Gene Accession Number                       AFFX-BioDn-5_at   
1                                                      -558   
2                                                      -400   
3                                                      -650   

                                                       6     \
Gene Description       AFFX-BioDn-3_at (endogenous control)   
Gene Accession Number                       AFFX-BioDn-3_at   
1                                                       199   
2                                                      -330   
3                                                        33   

                                                      7     \
Gene Description       AFFX-CreX-5_at (endogenous control)   
Gene Accession Number                       AFFX-CreX-5_at   
1                                                     -176   
2                                                     -168   
3                                                     -367   

                                                      8     \
Gene Description       AFFX-CreX-3_at (endogenous control)   
Gene Accession Number                       AFFX-CreX-3_at   
1                                                      252   
2                                                      101   
3                                                      206   

                                                      9     ...  \
Gene Description       AFFX-BioB-5_st (endogenous control)  ...   
Gene Accession Number                       AFFX-BioB-5_st  ...   
1                                                      206  ...   
2                                                       74  ...   
3                                                     -215  ...   

                                                            7119  \
Gene Description       Transcription factor Stat5b (stat5b) mRNA   
Gene Accession Number                                  U48730_at   
1                                                            185   
2         

Just clearning the data, removing extra columns and converting to numerical values.

In [13]:
# Clean up the column names for training and testing data
X_train.columns = X_train.iloc[1]
X_train = X_train.drop(["Gene Description", "Gene Accession Number"]).apply(pd.to_numeric)

# Clean up the column names for Testing data
X_test.columns = X_test.iloc[1]
X_test = X_test.drop(["Gene Description", "Gene Accession Number"]).apply(pd.to_numeric)

print(X_train.shape)
print(X_test.shape)
X_train.head()

(38, 7129)
(34, 7129)


Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \
1                                -214            -153             -58   
2                                -139             -73              -1   
3                                 -76             -49            -307   
4                                -135            -114             265   
5                                -106            -125             -76   

Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \
1                                  88            -295             -558   
2                                 283            -264             -400   
3                                 309            -376             -650   
4                                  12            -419             -585   
5                                 168            -230             -284   

Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \
1                                  199            -176             252   
2                                 -330            -168             101   
3                                   33            -367             206   
4                                  158            -253              49   
5                                    4            -122              70   

Gene Accession Number  AFFX-BioB-5_st  ...  U48730_at  U58516_at  U73738_at  \
1                                 206  ...        185        511       -125   
2                                  74  ...        169        837        -36   
3                                -215  ...        315       1199         33   
4                                  31  ...        240        835        218   
5                                 252  ...        156        649         57   

Gene Accession Number  X06956_at  X16699_at  X83863_at  Z17240_at  \
1                            389        -37        793        329   
2                            442        -17        782        295   
3                            168         52       1138        777   
4                            174       -110        627        170   
5                            504        -26        250        314   

Gene Accession Number  L49218_f_at  M71243_f_at  Z78285_f_at  
1                               36          191          -37  
2                               11           76          -14  
3                               41          228          -41  
4                              -50          126          -91  
5                               14           56          -25  

[5 rows x 7129 columns]

We have the 38 patients as rows in the training set, and the other 34 as rows in the testing set. Each of those datasets has 7129 gene expression features. But we haven't yet associated the target labels with the right patients. You will recall that all the labels are all stored in a single dataframe. Let's split the data so that the patients and labels match up across the training and testing dataframes.We are now splitting the data into train and test sets. We will subset the first 38 patient's cancer types.

In [14]:
X_train = X_train.reset_index(drop=True)
y_train = y[y.patient <= 38].reset_index(drop=True)

# Subset the rest for testing
X_test = X_test.reset_index(drop=True)
y_test = y[y.patient > 38].reset_index(drop=True)

Generate descriptive statistics to analyse the data further.

In [15]:
X_train.describe()

Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \
count                       38.000000       38.000000       38.000000   
mean                      -120.868421     -150.526316      -17.157895   
std                        109.555656       75.734507      117.686144   
min                       -476.000000     -327.000000     -307.000000   
25%                       -138.750000     -205.000000      -83.250000   
50%                       -106.500000     -141.500000      -43.500000   
75%                        -68.250000      -94.750000       47.250000   
max                         17.000000      -20.000000      265.000000   

Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \
count                       38.000000       38.000000        38.000000   
mean                       181.394737     -276.552632      -439.210526   
std                        117.468004      111.004431       135.458412   
min                        -36.000000     -541.000000      -790.000000   
25%                         81.250000     -374.250000      -547.000000   
50%                        200.000000     -263.000000      -426.500000   
75%                        279.250000     -188.750000      -344.750000   
max                        392.000000      -51.000000      -155.000000   

Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \
count                        38.000000       38.000000       38.000000   
mean                        -43.578947     -201.184211       99.052632   
std                         219.482393       90.838989       83.178397   
min                        -479.000000     -463.000000      -82.000000   
25%                        -169.000000     -239.250000       36.000000   
50%                         -33.500000     -185.500000       99.500000   
75%                          79.000000     -144.750000      152.250000   
max                         419.000000      -24.000000      283.000000   

Gene Accession Number  AFFX-BioB-5_st  ...   U48730_at    U58516_at  \
count                       38.000000  ...   38.000000    38.000000   
mean                       112.131579  ...  178.763158   750.842105   
std                        211.815597  ...   84.826830   298.008392   
min                       -215.000000  ...   30.000000   224.000000   
25%                        -47.000000  ...  120.000000   575.500000   
50%                         70.500000  ...  174.500000   700.000000   
75%                        242.750000  ...  231.750000   969.500000   
max                        561.000000  ...  356.000000  1653.000000   

Gene Accession Number   U73738_at    X06956_at   X16699_at    X83863_at  \
count                   38.000000    38.000000   38.000000    38.000000   
mean                     8.815789   399.131579  -20.052632   869.052632   
std                     77.108507   469.579868   42.346031   482.366461   
min                   -178.000000    36.000000 -112.000000   195.000000   
25%                    -42.750000   174.500000  -48.000000   595.250000   
50%                     10.500000   266.000000  -18.000000   744.500000   
75%                     57.000000   451.750000    9.250000  1112.000000   
max                    218.000000  2527.000000   52.000000  2315.000000   

Gene Accession Number    Z17240_at  L49218_f_at  M71243_f_at  Z78285_f_at  
count                    38.000000    38.000000    38.000000    38.000000  
mean                    335.842105    19.210526   504.394737   -29.210526  
std                     209.826766    31.158841   728.744405    30.851132  
min                      41.000000   -50.000000    -2.000000   -94.000000  
25%                     232.750000     8.000000   136.000000   -42.750000  
50%                     308.500000    20.000000   243.500000   -26.000000  
75%                     389.500000    30.250000   487.250000   -11.500000  
max                    1109.000000   115.000000  3193.000000    36.000000  

[8 rows

Clearly there is some variation in the scales across the different features. Many machine learning models work much better with data that's on the same scale, so let's create a scaled version of the dataset.

In [16]:
X_train_fl = X_train.astype(float, 64)
X_test_fl = X_test.astype(float, 64)

# Apply the same scaling to both datasets
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train_fl)
X_test = scaler.transform(X_test_fl) # note that we transform rather than fit_transform

In [21]:
y_test

patient  cancer
0        39       0
1        40       0
2        41       0
3        42       0
4        43       0
5        44       0
6        45       0
7        46       0
8        47       0
9        48       0
10       49       0
11       50       1
12       51       1
13       52       1
14       53       1
15       54       1
16       55       0
17       56       0
18       57       1
19       58       1
20       59       0
21       60       1
22       61       1
23       62       1
24       63       1
25       64       1
26       65       1
27       66       1
28       67       0
29       68       0
30       69       0
31       70       0
32       71       0
33       72       0

<a id='dask1'></a>

### 2. Conversion to CuDF Dataframe
Convert the pandas dataframes to CuDF dataframes to carry out the further CuML tasks.

In [22]:
#Modify the code in this cell

#%%time
X_cudf_train = cudf.DataFrame(X_train)  #Pass X train dataframe here
X_cudf_test = cudf.DataFrame(X_test)    #Pass X test dataframe here

y_cudf_train = cudf.DataFrame(y_train['cancer'].values)  #Pass y train dataframe here
y_cudf_test = cudf.Series(y_test['cancer'].values)   #Pass y test dataframe here

In [24]:
X_cudf_train

0         1         2         3         4         5         6     \
0  -0.861496 -0.033101 -0.351701 -0.805738 -0.168417 -0.888716  1.120068   
1  -0.167723  1.037400  0.139139  0.876572  0.114600  0.293351 -1.322502   
2   0.415047  1.358550 -2.495899  1.100880 -0.907912 -1.577008  0.353591   
3  -0.130721  0.488768  2.429729 -1.461407 -1.300484 -1.090715  0.930757   
4   0.137537  0.341574 -0.506703 -0.115559  0.425006  1.161198  0.219688   
5  -0.158472  0.876825  1.999167 -0.952401  0.041564 -0.888716  0.510580   
6   0.452048  0.087330  2.197225 -1.090436 -1.117893 -0.836346  0.806089   
7  -2.702307 -1.464896  0.208029 -1.582188 -2.414292 -2.624409 -1.068548   
8   1.164322  0.314812  1.060542  0.747163  0.607597 -0.716643  0.201219   
9   0.304043  0.609200  0.509423  0.324429  0.899744  1.445492  1.715704   
10 -0.408231 -0.059864 -0.463647 -0.857501  1.036687  0.068908  0.662952   
11  0.498299  0.769775  0.871095 -1.349253  0.890614  0.869422 -0.422121   
12  0.267042  0.421862 -0.119198 -0.072423  0.397617  1.587639 -0.025031   
13  0.072785  0.047187 -0.868375  0.531483  1.365352  0.308314 -0.948498   
14  0.128287  1.050781 -0.937265 -0.279477  0.653245  1.161198 -0.565259   
15  0.035784 -0.916264 -0.282811  0.652264  0.534561  0.278388  1.253971   
16 -3.285076 -0.835977 -0.007252  1.031862 -1.154411  0.338240  0.007291   
17  0.368795  0.007043 -0.876987 -0.892010  1.137113  0.742237  0.034995   
18  0.711056  1.331788  1.008875  0.220902  1.191890  1.632528  0.584457   
19  1.275325 -1.050077  0.828039  0.315802  0.132859  0.263425  1.706469   
20 -0.213974 -0.648639 -1.204214 -0.426140  1.146242  0.689867 -0.343627   
21 -1.166756  0.809919 -1.298937 -1.771988 -0.287101 -0.985974 -0.583728   
22  0.433547 -2.281153  0.053027 -1.875514 -0.369267 -0.447311 -0.435973   
23  0.008033 -1.505040 -0.833930  0.635009 -0.597507  0.323277 -1.701123   
24  0.368795  0.007043 -0.584205  1.161270 -1.291355 -0.163017 -0.103525   
25  0.082036 -1.103602 -0.523926 -1.099064  0.297192  1.228531 -2.010484   
26 -1.407264 -2.361441 -0.506703 -0.866128 -1.483076 -1.322639  2.135881   
27  1.081069  0.462006 -0.928654  0.514228  0.781059  0.211055  0.058081   
28  1.256825  0.488768  0.164973  0.100121  2.059200  2.126303  0.335121   
29 -1.823528 -0.554970 -0.670317  1.126762  1.255797  0.712311  1.697235   
30  0.822060  1.358550  0.569701  0.419329 -0.825746 -0.514644 -1.410231   
31 -0.028968  0.957113 -0.170865  1.282052  0.808448  0.121278  0.058081   
32 -0.130721 -0.474683 -0.455036  1.342442 -1.190929 -0.948567 -0.449825   
33  0.933064 -0.755689 -0.282811 -0.693584 -0.844005 -0.671754  0.265861   
34  1.182822  0.676106 -0.343090 -0.426140 -0.917042 -0.290201 -1.419466   
35 -0.852245 -1.357846  1.318879  1.178525  0.616727 -0.881234  0.385912   
36  0.886812  1.746607  1.215544  1.238916 -1.090504 -0.185461 -0.819212   
37  0.452048  0.154237  0.139139  1.816940 -0.433174 -0.529607 -1.414848   

        7         8         9     ...      7119      7120      7121      7122  \
0   0.280962  1.863470  0.449110  ...  0.074511 -0.815620 -1.758717 -0.021865   
1   0.370212  0.023726 -0.182439  ... -0.116640  0.292993 -0.589006  0.092516   
2  -1.849884  1.303018 -1.565148  ...  1.627617  1.524030  0.317849 -0.498816   
3  -0.578070 -0.609828 -0.388171  ...  0.731595  0.286192  2.749271 -0.485868   
4   0.883400 -0.353970  0.669195  ... -0.271951 -0.346330  0.633277  0.226322   
5   0.169399 -0.146846  0.386912  ... -0.761776  1.598844 -1.114719 -0.490184   
6   0.247493  0.328320 -0.632178  ... -1.777269  0.231781 -2.455287 -0.535505   
7  -2.920885 -0.353970 -1.345063  ...  1.316996 -0.414343 -1.246147 -0.209624   
8   0.303274 -0.914421  1.884448  ...  2.117443  0.779287 -0.037007 -0.479393   
9   0.593337  0.949690  0.339067  ... -1.633905  0.799691  0.225849 -0.643412   
10  1.028431 -0.524542  1.138072  ...  0.074511 -0.370134  0.304706 -0.565719   
11  0.827618 -1.231198 -0.852264  ... -1.562223 -1.791608  0.672705 -0.

### 3. Model Building
#### Dask Integration

We will try using the Random Forests Classifier  and implement using CuML and Dask.

#### Start Dask cluster

In [25]:
#Modify the code in this cell

# This will use all GPUs on the local host by default
cluster = LocalCUDACluster(threads_per_worker=1) #Set 1 thread per worker using arguments to cluster
c = Client(cluster) #Pass the cluster as an argument to Client

# Query the client for all connected workers
workers = c.has_what().keys()
n_workers = len(workers)
n_streams = 8 # Performance optimization

NVMLError_NotFound: Not Found

In [26]:
# This will use all GPUs on the local host by default
cluster = LocalCUDACluster(threads_per_worker=1)
c = Client(cluster)

# Query the client for all connected workers
workers = c.has_what().keys()
n_workers = len(workers)
n_streams = 8 # Performance optimization

NVMLError_NotFound: Not Found

#### Define Parameters

In addition to the number of examples, random forest fitting performance depends heavily on the number of columns in a dataset and (especially) on the maximum depth to which trees are allowed to grow. Lower `max_depth` values can greatly speed up fitting, though going too low may reduce accuracy.

In [28]:
# Random Forest building parameters
max_depth = 12
n_bins = 16
n_trees = 1000

#### Distribute data to worker GPUs

In [29]:
X_train = X_train.astype(np.float32)
X_test = X_test.astype(np.float32)
y_train = y_train.astype(np.int32)
y_test = y_test.astype(np.int32)

In [ ]:
n_partitions = n_workers

def distribute(X, y):
    # First convert to cudf (with real data, you would likely load in cuDF format to start)
    X_cudf = cudf.DataFrame.from_pandas(pd.DataFrame(X))
    y_cudf = cudf.Series(y)

    # Partition with Dask
    # In this case, each worker will train on 1/n_partitions fraction of the data
    X_dask = dask_cudf.from_cudf(X_cudf, npartitions=n_partitions)
    y_dask = dask_cudf.from_cudf(y_cudf, npartitions=n_partitions)

    # Persist to cache the data in active memory
    X_dask, y_dask = \
      dask_utils.persist_across_workers(c, [X_dask, y_dask], workers=workers)
    
    return X_dask, y_dask

In [ ]:
#Modify the code in this cell

X_train_dask, y_train_dask = distribute(X_train, y_train) #Pass train data as arguments here
X_test_dask, y_test_dask = distribute(X_test, y_test) #Pass test data as arguments here

#### Create the  Scikit-learn model

Since a scikit-learn equivalent to the multi-node multi-GPU K-means in cuML doesn't exist, we will use Dask-ML's implementation for comparison.

In [59]:
%%time
max_depth = 12
n_trees = 150

# Use all avilable CPU cores
skl_model = sklRF(n_estimators=n_trees, max_depth=max_depth, n_jobs=-1)
skl_model.fit(X_train, y_train.iloc[:,1])

CPU times: user 206 ms, sys: 16.8 ms, total: 223 ms
Wall time: 242 ms


RandomForestClassifier(max_depth=12, n_estimators=150, n_jobs=-1)

In [63]:
import xgboost as xgb
xgr=xgb.XGBClassifier(max_depth=50,min_child_weight=0.1,gamma=0.0001)
dtrain = xgb.DMatrix(X_train, label=y_train.iloc[:,1])
dtest = xgb.DMatrix(X_test, label=y_test.iloc[:,1])



In [64]:
# instantiate params
params = {}

# general params
general_params = {}
params.update(general_params)

# booster params
booster_params = {'tree_method': 'gpu_hist'}
params.update(booster_params)

# learning task params
learning_task_params = {'objective': 'reg:squarederror'}
params.update(learning_task_params)
print(params)

{'tree_method': 'gpu_hist', 'objective': 'reg:squarederror'}


In [65]:
evallist = [(dtest, 'test'), (dtrain, 'train')]
num_round = 300

In [66]:
bst = xgb.train(params, dtrain, num_round, evallist)


[0]	test-rmse:0.39144	train-rmse:0.35744
[1]	test-rmse:0.32951	train-rmse:0.25557
[2]	test-rmse:0.29880	train-rmse:0.18276
[3]	test-rmse:0.28634	train-rmse:0.13072
[4]	test-rmse:0.28309	train-rmse:0.09351
[5]	test-rmse:0.28381	train-rmse:0.06691
[6]	test-rmse:0.28591	train-rmse:0.04788
[7]	test-rmse:0.28821	train-rmse:0.03427
[8]	test-rmse:0.29027	train-rmse:0.02453
[9]	test-rmse:0.29195	train-rmse:0.01757
[10]	test-rmse:0.29327	train-rmse:0.01258
[11]	test-rmse:0.29427	train-rmse:0.00901
[12]	test-rmse:0.29502	train-rmse:0.00645
[13]	test-rmse:0.29557	train-rmse:0.00462
[14]	test-rmse:0.29597	train-rmse:0.00331
[15]	test-rmse:0.29627	train-rmse:0.00238
[16]	test-rmse:0.29648	train-rmse:0.00170
[17]	test-rmse:0.29664	train-rmse:0.00122
[18]	test-rmse:0.29675	train-rmse:0.00088
[19]	test-rmse:0.29683	train-rmse:0.00063
[20]	test-rmse:0.29689	train-rmse:0.00045
[21]	test-rmse:0.29693	train-rmse:0.00032
[22]	test-rmse:0.29696	train-rmse:0.00023
[23]	test-rmse:0.29699	train-rmse:0.00017
[2

In [82]:
y_pred_train = bst.predict(dtrain)
y_pred_train = np.round(y_pred_train).astype(np.int32)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0], dtype=int32)

In [84]:
print("XGboost accuracy:  ", accuracy_score(y_train.iloc[:,1], y_pred_train))


XGboost accuracy:   1.0


In [85]:
y_pred_test = bst.predict(dtest)
y_pred_test = np.round(y_pred_test).astype(np.int32)
print("XGboost accuracy:  ", accuracy_score(y_test.iloc[:,1], y_pred_test))

XGboost accuracy:   0.9117647058823529


In [73]:
y_test.iloc[:,1]

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    1
12    1
13    1
14    1
15    1
16    0
17    0
18    1
19    1
20    0
21    1
22    1
23    1
24    1
25    1
26    1
27    1
28    0
29    0
30    0
31    0
32    0
33    0
Name: cancer, dtype: int32

In [71]:
import cupy as cp
from cuml.metrics.regression import r2_score

y_pred_test = bst.predict(dtest)

y_pred_cp = cp.asarray(y_pred_test)
y_test_cp = cp.asarray(y_train).astype(np.float32)

print("SKLearn accuracy:  ", accuracy_score(y_test.iloc[:,1], y_pred_cp))

TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.

In [34]:
from sklearn.model_selection import GridSearchCV
params_grid = {'max_depth': [10, 15, 50, 100], 'n_estimators': [500, 1000, 1500, 2000]}
grid = GridSearchCV(sklRF(), params_grid, scoring='accuracy')
grid.fit(X_train, y_train.iloc[:,1])

GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'max_depth': [10, 15, 50, 100],
                         'n_estimators': [500, 1000, 1500, 2000]},
             scoring='accuracy')

In [35]:
grid.best_params_

{'max_depth': 10, 'n_estimators': 500}

#### Train the distributed cuML model

In [31]:
#Modify the code in this cell

%%time

cuml_model = cumlDaskRF(max_depth=max_depth, n_estimators=n_trees, n_bins=n_bins, n_streams=n_streams)
cuml_model.fit() # Pass X and y train dask data here

wait(cuml_model.rfs) # Allow asynchronous training tasks to finish

UsageError: Line magic function `%%time` not found.


#### Predict and check accuracy

In [74]:
#Modify the code in this cell

skl_y_pred = skl_model.predict(X_train)
#cuml_y_pred = cuml_model.predict().compute().to_array()  #Pass the X test dask data as argument here

# Due to randomness in the algorithm, you may see slight variation in accuracies
print("SKLearn accuracy:  ", accuracy_score(y_train.iloc[:,1], skl_y_pred))
#print("CuML accuracy:     ", accuracy_score())  #Pass the y test dask data  and predicted values from CuML model as argument here

SKLearn accuracy:   1.0


In [75]:
skl_y_pred

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

<a id='ex4'></a><br>

### 4. CONCLUSION

Let's compare the performance of our solution!

| Algorithm     | Implementation | Accuracy      | Time | Algorithm     | Implementation | Accuracy      | Time |
| ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |


Write down your observations and compare the CuML and Scikit learn scores. They should be approximately equal.  We hope that you found this exercise exciting and beneficial in understanding RAPIDS better. Share your highest accuracy and try to use the unique features of RAPIDS for accelerating your data science pipelines. Don't restrict yourself to the previously explained concepts, but use the documentation to apply more models and functions and achieve the best results. Jump over to the next notebook for our sample solution.


### 5.  References



<p xmlns:dct="http://purl.org/dc/terms/">
  <a rel="license"
     href="http://creativecommons.org/publicdomain/zero/1.0/">
    <center><img src="http://i.creativecommons.org/p/zero/1.0/88x31.png" style="border-style: none;" alt="CC0"  /></center>
  </a>
 
</p>


- The dataset is licensed under a CC0: Public Domain license.

- Molecular Classification of Cancer: Class Discovery and Class Prediction by Gene Expression. Science 286:531-537. (1999). Published: 1999.10.14. T.R. Golub, D.K. Slonim, P. Tamayo, C. Huard, M. Gaasenbeek, J.P. Mesirov, H. Coller, M. Loh, J.R. Downing, M.A. Caligiuri, C.D. Bloomfield, and E.S. Lander


## Licensing
  
This material is released by OpenACC-Standard.org, in collaboration with NVIDIA Corporation, under the Creative Commons Attribution 4.0 International (CC BY 4.0).

&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;&emsp;&emsp;
&emsp;&emsp;&emsp;
[Home Page](../../START_HERE.ipynb)
